src/platform_impl/macos/mod.rs

```diff
@@ -120,79 +112,87 @@ impl TrayIcon {
     }
 
     pub fn set_icon(&mut self, icon: Option<Icon>) -> crate::Result<()> {
-        if let (Some(ns_status_item), Some(tray_target)) = (self.ns_status_item, self.tray_target) {
-            set_icon_for_ns_status_item_button(ns_status_item, icon.clone(), false)?;
-            unsafe {
-                let _: () = msg_send![tray_target, updateDimensions];
-            }
+        if let (Some(ns_status_item), Some(tray_target)) = (&self.ns_status_item, &self.tray_target)
+        {
+            set_icon_for_ns_status_item_button(&ns_status_item, icon.clone(), false, self.mtm)?;
+            tray_target.update_dimensions();
         }
         self.attrs.icon = icon;
         Ok(())
     }
 
     pub fn set_menu(&mut self, menu: Option<Box<dyn menu::ContextMenu>>) {
-        if let (Some(ns_status_item), Some(tray_target)) = (self.ns_status_item, self.tray_target) {
+        if let (Some(ns_status_item), Some(tray_target)) = (&self.ns_status_item, &self.tray_target)
+        {
             unsafe {
-                let menu = menu.as_ref().map(|m| m.ns_menu() as _).unwrap_or(nil);
-                (*tray_target).set_ivar(TRAY_MENU, menu);
-                ns_status_item.setMenu_(menu);
-                let () = msg_send![menu, setDelegate: ns_status_item];
+                let menu = menu
+                    .as_ref()
+                    .and_then(|m| m.ns_menu().cast::<NSMenu>().as_ref())
+                    .map(|menu| menu.retain());
+                ns_status_item.setMenu(menu.as_deref());
+                if let Some(menu) = &menu {
+                    let () = msg_send![menu, setDelegate: &**ns_status_item];
```

Comment:
I'm not sure what the original code intended here, but this is wrong, `NSStatusItem` doesn't implement `NSMenuDelegate` (at least not publicly).

Reply:
Hmm, it was set to `tray_target`. Maybe @lucasfernog has some ideas?